evm_arithmetization/src/cpu/kernel/asm/smt/delete.asm

```diff
@@ -0,0 +1,251 @@
+    DUP1 %eq_const(1) %jumpi(smt_delete_internal_1)
+    DUP1 %eq_const(2) %jumpi(smt_delete_internal_2)
+    DUP1 %eq_const(3) %jumpi(smt_delete_internal_3)
+    PANIC
+smt_delete_internal_0:
+    // stack: level%4, node_payload_ptr, level, ks, retdest
+    %stack (level_mod_4, node_payload_ptr, level, k0, k1, k2, k3 ) -> (k0, node_payload_ptr, level, k0, k1, k2, k3 )
+    %pop_bit
+    %stack (bit, newk0, node_payload_ptr, level, k0, k1, k2, k3 ) -> (bit, node_payload_ptr, level, newk0, k1, k2, k3 )
+    %jump(smt_delete_internal_contd)
+smt_delete_internal_1:
+    %stack (level_mod_4, node_payload_ptr, level, k0, k1, k2, k3 ) -> (k1, node_payload_ptr, level, k0, k1, k2, k3 )
+    %pop_bit
+    %stack (bit, newk1, node_payload_ptr, level, k0, k1, k2, k3 ) -> (bit, node_payload_ptr, level, k0, newk1, k2, k3 )
+    %jump(smt_delete_internal_contd)
+smt_delete_internal_2:
+    %stack (level_mod_4, node_payload_ptr, level, k0, k1, k2, k3 ) -> (k2, node_payload_ptr, level, k0, k1, k2, k3 )
+    %pop_bit
+    %stack (bit, newk2, node_payload_ptr, level, k0, k1, k2, k3 ) -> (bit, node_payload_ptr, level, k0, k1, newk2, k3 )
+    %jump(smt_delete_internal_contd)
+smt_delete_internal_3:
+    %stack (level_mod_4, node_payload_ptr, level, k0, k1, k2, k3 ) -> (k3, node_payload_ptr, level, k0, k1, k2, k3 )
+    %pop_bit
+    %stack (bit, newk3, node_payload_ptr, level, k0, k1, k2, k3 ) -> (bit, node_payload_ptr, level, k0, k1, k2, newk3 )
+smt_delete_internal_contd:
+    //stack: bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    PUSH internal_update
+    //stack: internal_update, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    %rep 7
+        DUP8
+    %endrep
+    //stack: bit, node_payload_ptr, level, k0, k1, k2, k3, internal_update, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    ADD
+    //stack: child_ptr_ptr, level, k0, k1, k2, k3, internal_update, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    %mload_trie_data
+    //stack: child_ptr, level, k0, k1, k2, k3, internal_update, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    SWAP1 %increment SWAP1
+    //stack: child_ptr, level+1, k0, k1, k2, k3, internal_update, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    %jump(smt_delete_with_keys)
+
+// Update the internal node, possibly deleting it, or returning a leaf node.
+internal_update:
+    // Update the child first.
+    //stack: deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP2 PUSH 1 SUB
+    //stack: 1-bit, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP4 ADD
+    //stack: sibling_ptr_ptr, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %mload_trie_data DUP1 %mload_trie_data
+    //stack: sibling_node_type, sibling_ptr, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP1 %eq_const(@SMT_NODE_HASH) %jumpi(sibling_is_hash)
+    %eq_const(@SMT_NODE_LEAF) %jumpi(sibling_is_leaf)
+sibling_is_internal:
```

Comment:
A bit surprised to not see the "panic if none of the options" logic here. Do we have better guarantees here than in the other places?

Reply:
Added a PANIC